src/compiler/checker.ts

```diff
@@ -23720,6 +23828,12 @@ namespace ts {
             return type.flags & TypeFlags.Union ? forEach((type as UnionType).types, f) : f(type);
         }
 
+        function reduceType<T>(type: Type, f: (memo: T, t: Type) => T | undefined, initial: T): T;

```

Comment:
Not sure this function is really needed per my comment above.

Reply:
I'm actually surprised we haven't added it before this. I know I've had a occasions where I thought it would be useful. Functions like `maybeTypeOfKind` essentially manually unroll what `reduceType`, and there are more than a few other places in the checker that could be rewritten to use this function as well.